subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/generator/TestProjectGeneratorConfiguration.groovy

```diff
@@ -41,5 +50,89 @@ class TestProjectGeneratorConfiguration {
     int maxParallelForks
     int testForkEvery
     boolean useTestNG
-    Map<String, String> fileToChangeByScenario = [:]
+    Map<String, String> fileToChangeByScenario
+}
+
+@Builder(prefix = "with",
+    builderStrategy = ExternalStrategy,
+    forClass = TestProjectGeneratorConfiguration,
+    excludes = [
+        'plugins',
+        'repositories',
+        'externalApiDependencies',
+        'externalImplementationDependencies',
+        'minLinesOfCodePerSourceFile',
+        'testRunnerMemory',
+        'parallel',
+        'maxWorkers',
+        'testForkEvery',
+        'maxParallelForks'
+    ])
+class TestProjectGeneratorConfigurationBuilder {
+    TestProjectGeneratorConfigurationBuilder(String projectName, Language language = Language.JAVA) {
+        this(projectName, projectName, language)
+    }
+
+    TestProjectGeneratorConfigurationBuilder(String projectName, String templateName, Language language = Language.JAVA) {
+        this.projectName = projectName
+        this.templateName = templateName
+        this.language = language
```

Comment:
Should we provide `language` via the builder as well? What things do we need to pass to the constructor and what via the builder?

Reply:
It's intentional to make `language` a constructor parameter: 

```
    private String productionFile(int project = 0, int pkg = 0, int file = 0) {
        if (project >= 0) {
            "project${project}/src/main/${this.language.name}/org/gradle/test/performance/${this.templateName.toLowerCase()}/project${project}/p${pkg}/Production${file}.${this.language.name}"
        } else {
            "src/main/${this.language.name}/org/gradle/test/performance/${this.templateName.toLowerCase()}/p${pkg}/Production${file}.${this.language.name}"
        }
    }
```

This method depends on `language` so we must initialize `language` in constructor and not change it any more.

Reply:
It surprises me that the (mutable) builder doesn't finalize all its state in the `create` method and requires some immutable state already.